masonry/src/passes/layout.rs

```diff
@@ -25,207 +25,139 @@ fn rect_contains(larger: &Rect, smaller: &Rect) -> bool {
-    let id = pod.id().to_raw();
-    let widget_mut = parent_ctx
-        .widget_children
-        .get_child_mut(id)
-        .expect("WidgetPod: inner widget not found in widget tree");
-    let mut state_mut = parent_ctx
-        .widget_state_children
-        .get_child_mut(id)
-        .expect("WidgetPod: inner widget not found in widget tree");
-    let widget = widget_mut.item;
-    let state = state_mut.item;
-
-    // The parent (and only the parent) controls the stashed state, and it is valid to set `stashed` in layout.
-    // Because of that, we use the local value rather than the global value.
-    // Note that if we are stashed by a grandparent, this check would trigger for that grandparent, so we should
-    // never be called.
-    if state.is_explicitly_stashed {
+    // This checks reads is_explicitly_stashed instead of is_stashed because the latter may be outdated.
+    // A widget's is_explicitly_stashed flag is controlled by its direct parent.
+    // The parent may set this flag during layout, in which case it should avoid calling run_layout.
+    // Note that, because this check exits before recursing, run_layout can only ever be
+    // reached for a widget whose parent is not stashed, which means is_explicitly_stashed
+    // being false is sufficient to know the widget is non-stashed.
+    if state.item.is_explicitly_stashed {
         debug_panic!(
             "Error in '{}' {}: trying to compute layout of stashed widget.",
-            widget.short_type_name(),
+            widget.item.short_type_name(),
             pod.id(),
         );
-        state.size = Size::ZERO;
-        return false;
+        state.item.size = Size::ZERO;
+        return Size::ZERO;
     }
 
     // TODO - Not everything that has been re-laid out needs to be repainted.
-    state.needs_paint = true;
-    state.needs_compose = true;
-    state.needs_accessibility = true;
-    state.request_paint = true;
-    state.request_compose = true;
-    state.request_accessibility = true;
-
-    bc.debug_check(widget.short_type_name());
+    state.item.needs_paint = true;
+    state.item.needs_compose = true;
+    state.item.needs_accessibility = true;
+    state.item.request_paint = true;
+    state.item.request_compose = true;
+    state.item.request_accessibility = true;
+
+    bc.debug_check(widget.item.short_type_name());
     trace!("Computing layout with constraints {:?}", bc);
 
-    state.local_paint_rect = Rect::ZERO;
+    state.item.local_paint_rect = Rect::ZERO;
+
+    // TODO - Handle more elegantly
+    // We suppress need_layout and request_layout for stashed children
+    // to avoid unnecessary relayouts in corner cases.
+    recurse_on_children(
+        pod.id(),
+        widget.reborrow_mut(),
+        state.children.reborrow_mut(),
+        |_, state| {
+            if state.item.is_stashed {
+                state.item.needs_layout = false;
+                state.item.request_layout = false;
+            }
+        },
+    );
```

Comment:
Five minutes after pushing this code, it occurs to me it basically doesn't work as a safety net, because it doesn't cover grandchildren.

I really need to sit down and write down how layout and stashing interact at some point.